lib/housekeeping.js

```diff
@@ -38,17 +38,109 @@ class HouseKeeper {
+    //       ${vol-type}: {
+    //         active: {
+    //           gb: value,
+    //           count: value
+    //         },
+    //         unused: {
+    //           gb: value,
+    //           count: value
+    //         }
+    //       }
+    //     },
+    //     ...
+    //   ]
+    // }
+
+    //assert(typeof totalsByCategory === 'object');
+
+    //assert(Object.keys(totalsByCategory).length === 1);
+    let region =  Object.keys(totalsByCategory)[0];
+    
+    let volumeTypes = totalsByCategory[region];
+    //assert(typeof volumeTypes === 'object');
+    //assert(typeof volumeTypes.hasOwnProperty('length'));
+
+    volumeTypes.forEach(function(data) {
+      //assert(typeof data === 'object');
+      
+      //assert(Object.keys(data).length === 1);
+      let type = Object.keys(data)[0];
+
+      //assert(data.hasOwnProperty('active'));
+      //assert(data.hasOwnProperty('unused'));
+      let active = data.active;
+      let unused = data.unused;
+
+      //assert(typeof active === 'object');
+      //assert(active.hasOwnProperty('gb'));
+      //assert(active.hasOwnProperty('count'));
+
+      this.monitor.count('ebs-volumes:${region}:gb-usage', active.gb);
+      this.monitor.count('ebs-volumes:${region}:count', active.count);
+
+      //assert(typeof unused === 'object');
+      //assert(unused.hasOwnProperty('gb'));
+      //assert(unused.hasOwnProperty('count'));
+
+      this.monitor.count('ebs-volumes:${region}:unused-gb-usage', unused.gb);
+      this.monitor.count('ebs-volumes:${region}:unused-count', unused.count);
+
+      let totalGbUsage = active.gb + unused.gb;
+      let totalCount = active.count + unused.count;
+      this.monitor.count('ebs-volumes:global:gb-usage', totalGbUsage);
+      this.monitor.count('ebs-volumes:global:count', totalCount);
+      this.monitor.count('ebs-volumes:${region}:type:${type}:gb-usage', totalGbUsage);
+      this.monitor.count('ebs-volumes:${region}:type:${type}:count', totalCount);
+    })
+}
+
+  _handleVolumeData(volumes) {
+    // First, calculate total GBs and volumes by category
+    let totals = this._calculateVolumeTotals(volumes);
+
+    // Now, send an update to each corresponding statsum series with current totals.
+    this._submitVolumeTotals(totals);
+  }
 
   async _sweepVolumes(region) {
-    let volumes = await this.runaws(this.ec2[region], 'describeVolumes', {
-      Filters: [{
-        Name: 'status',
-        Values: ['available', 'in-use'],
-      }],
-    });
-
-    await Promise.all(volumes.Volumes.map(this._handleVolumeData));
+    // Collect all volume objects returned from AWS in a single array
+    let volumes = []
+
+    // We need to keep track of the NextToken value returned by AWS. Its value is non-null
+    // if the number of entries in AWS exceeds MaxResults and the next page can be retrieved
+    // by specifying NextToken in a subsequent request.
+    let hasNextToken = true;
+    let nextToken = null;
+    while (hasNextToken) {
```

Comment:
Maybe get rid of `hasNextToken` and change this loop to a do/while with the condition `nextToken != null`.

Reply:
That's a great idea! I forgot that such a thing existed.

Reply:
I don't think every language has do/while loops so it's not very popular. In "The Art of Readable Code", they compare it to going down a ski slope with the warning sign at the bottom (https://www.safaribooksonline.com/library/view/the-art-of/9781449318482/ch07s04.html).